File: clients/healthKit/healthKit/Services/SyncLedger.swift
import Foundation
import OSLog

/// File-based persistence for sync payloads and statistics.
///
/// Stores:
/// - Cumulative stats (record counts, breakdowns by type)
/// - Last NDJSON payload per record type (for demo verification)
/// - Recent sync events log (last 20, without payloads)
///
/// Thread-safe via actor isolation. Files live in the app's Documents directory
/// under `sync_ledger/`.
actor SyncLedger {

    private let baseDirectory: URL
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder

    private var stats: SyncStats
    private var recentEvents: [SyncRecord]

    private static let maxRecentEvents = 20

    init() {
        let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
        self.baseDirectory = docs.appendingPathComponent("sync_ledger", isDirectory: true)
        try? FileManager.default.createDirectory(at: baseDirectory, withIntermediateDirectories: true)

        let enc = JSONEncoder()
        enc.dateEncodingStrategy = .iso8601
        enc.outputFormatting = [.prettyPrinted, .sortedKeys]
        self.encoder = enc

        let dec = JSONDecoder()
        dec.dateDecodingStrategy = .iso8601
        self.decoder = dec

        // Load persisted state
        self.stats = Self.loadJSON(SyncStats.self, from: baseDirectory.appendingPathComponent("stats.json"), decoder: dec)
            ?? .empty
        self.recentEvents = Self.loadJSON([SyncRecord].self, from: baseDirectory.appendingPathComponent("recent_events.json"), decoder: dec)
            ?? []
    }

    // MARK: - Public API

    /// Record a successful sync POST.
    func recordSync(
        recordType: String,
        recordCount: Int,
        httpStatusCode: Int,
        ndjsonPayload: String,
        requestHeaders: [String: String]
    ) {
        let record = SyncRecord(
            id: UUID(),
            recordType: recordType,
            timestamp: Date(),
            recordCount: recordCount,
            httpStatusCode: httpStatusCode,
            success: true,
            ndjsonPayload: ndjsonPayload,
            requestHeaders: requestHeaders
        )

        // Update cumulative stats
        stats.totalRecordsSent[recordType, default: 0] += recordCount
        stats.lastSyncTimestamp[recordType] = record.timestamp
        updateBreakdowns(recordType: recordType, ndjsonPayload: ndjsonPayload, recordCount: recordCount)

        // Save last payload for this record type
        let payloadFile = baseDirectory.appendingPathComponent("last_payload_\(recordType).json")
        saveJSON(record, to: payloadFile)

        // Append to recent events (without payload to save space)
        let eventRecord = SyncRecord(
            id: record.id,
            recordType: record.recordType,
            timestamp: record.timestamp,
            recordCount: record.recordCount,
            httpStatusCode: record.httpStatusCode,
            success: record.success,
            ndjsonPayload: nil,
            requestHeaders: record.requestHeaders
        )
        recentEvents.insert(eventRecord, at: 0)
        if recentEvents.count > Self.maxRecentEvents {
            recentEvents = Array(recentEvents.prefix(Self.maxRecentEvents))
        }

        // Persist
        saveJSON(stats, to: baseDirectory.appendingPathComponent("stats.json"))
        saveJSON(recentEvents, to: baseDirectory.appendingPathComponent("recent_events.json"))
    }

    /// Current cumulative stats.
    func getStats() -> SyncStats {
        stats
    }

    /// Recent sync events (most recent first, no payload).
    func getRecentEvents() -> [SyncRecord] {
        recentEvents
    }

    /// Last successful payload for a given record type.
    func getLastPayload(for recordType: String) -> SyncRecord? {
        let file = baseDirectory.appendingPathComponent("last_payload_\(recordType).json")
        return Self.loadJSON(SyncRecord.self, from: file, decoder: decoder)
    }

    // MARK: - Breakdown Parsing

    /// Parse NDJSON lines to extract type fields for per-category breakdowns.
    private func updateBreakdowns(recordType: String, ndjsonPayload: String, recordCount: Int) {
        let lines = ndjsonPayload.split(separator: "\n")

        switch recordType {
        case "samples":
            for line in lines {
                if let type = extractField("type", from: line) {
                    stats.sampleBreakdown[type, default: 0] += 1
                }
            }
        case "workouts":
            for line in lines {
                if let activityType = extractField("activity_type", from: line) {
                    stats.workoutBreakdown[activityType, default: 0] += 1
                }
            }
        case "sleep":
            stats.sleepSessionCount += recordCount
        case "activity_summaries":
            stats.activitySummaryDayCount += recordCount
        case "deletes":
            for line in lines {
                if let sampleType = extractField("sample_type", from: line) {
                    stats.deleteBreakdown[sampleType, default: 0] += 1
                }
            }
        default:
            break
        }
    }

    /// Extract a string field value from a single JSON line.
    private func extractField(_ key: String, from line: Substring) -> String? {
        guard let data = line.data(using: .utf8),
              let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let value = dict[key] as? String else {
            return nil
        }
        return value
    }

    // MARK: - File I/O

    private func saveJSON<T: Encodable>(_ value: T, to url: URL) {
        do {
            let data = try encoder.encode(value)
            try data.write(to: url, options: .atomic)
        } catch {
            Log.sync.error("SyncLedger: failed to write \(url.lastPathComponent) — \(error.localizedDescription)")
        }
    }

    private static func loadJSON<T: Decodable>(_ type: T.Type, from url: URL, decoder: JSONDecoder) -> T? {
        guard let data = try? Data(contentsOf: url) else { return nil }
        return try? decoder.decode(type, from: data)
    }
}

File: androidHealthConnect/app/src/main/java/com/dbxwearables/android/data/repository/SyncLedger.kt
package com.dbxwearables.android.data.repository

import android.content.Context
import android.util.Log
import com.dbxwearables.android.data.model.SyncRecord
import com.dbxwearables.android.data.model.SyncStats
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*
import java.io.File
import java.time.Instant
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

/**
 * File-based persistence for sync payloads and cumulative statistics.
 *
 * Stores:
 * - Cumulative stats (record counts, breakdowns by type) in `stats.json`
 * - Last NDJSON payload per record type in `last_payload_{type}.json`
 * - Recent sync events (last 20, without payloads) in `recent_events.json`
 *
 * Thread-safe via [Mutex]. Files live in the app's internal storage under `sync_ledger/`.
 * This is the Android equivalent of the iOS SyncLedger actor.
 */
@Singleton
class SyncLedger @Inject constructor(
    @ApplicationContext context: Context
) {

    private val ledgerDir: File =
        File(context.filesDir, "sync_ledger").also { it.mkdirs() }

    private val json = Json {
        prettyPrint = true
        ignoreUnknownKeys = true
        encodeDefaults = true
    }

    private val mutex = Mutex()

    companion object {
        private const val TAG = "SyncLedger"
        private const val MAX_RECENT_EVENTS = 20
        private const val STATS_FILE = "stats.json"
        private const val RECENT_EVENTS_FILE = "recent_events.json"
    }

    // ---- Public API ----

    /**
     * Record a sync POST result: updates cumulative stats, saves the last payload
     * for the given [recordType], and appends to the recent events list.
     */
    suspend fun recordSync(
        recordType: String,
        recordCount: Int,
        httpStatusCode: Int,
        success: Boolean,
        ndjsonPayload: String?,
        requestHeaders: Map<String, String>
    ) = mutex.withLock {
        val record = SyncRecord(
            id = UUID.randomUUID().toString(),
            recordType = recordType,
            timestamp = Instant.now().toString(),
            recordCount = recordCount,
            httpStatusCode = httpStatusCode,
            success = success,
            ndjsonPayload = ndjsonPayload,
            requestHeaders = requestHeaders
        )

        // Load and update cumulative stats
        val stats = loadStats()
        stats.totalRecordsSent[recordType] =
            (stats.totalRecordsSent[recordType] ?: 0) + recordCount
        stats.lastSyncTimestamp[recordType] = record.timestamp
        updateBreakdowns(stats, recordType, ndjsonPayload, recordCount)
        saveFile(STATS_FILE, json.encodeToString(stats))

        // Save full record (with payload) as last payload for this record type
        saveFile(
            "last_payload_${recordType}.json",
            json.encodeToString(record)
        )

        // Append to recent events (without payload to save space)
        val recentEvents = loadRecentEvents().toMutableList()
        val eventRecord = record.copy(ndjsonPayload = null)
        recentEvents.add(0, eventRecord)
        val trimmed = if (recentEvents.size > MAX_RECENT_EVENTS) {
            recentEvents.take(MAX_RECENT_EVENTS)
        } else {
            recentEvents
        }
        saveFile(RECENT_EVENTS_FILE, json.encodeToString(trimmed))
    }

    /**
     * Returns cumulative sync stats, or [SyncStats.empty] if none persisted.
     */
    suspend fun getStats(): SyncStats = mutex.withLock {
        loadStats()
    }

    /**
     * Returns the last payload record for [recordType], or null if none stored.
     */
    suspend fun getLastPayload(recordType: String): SyncRecord? = mutex.withLock {
        loadFile<SyncRecord>("last_payload_${recordType}.json")
    }

    /**
     * Returns the list of recent sync events (most recent first, without payloads).
     */
    suspend fun getRecentEvents(): List<SyncRecord> = mutex.withLock {
        loadRecentEvents()
    }

    // ---- Breakdown Parsing ----

    /**
     * Parse NDJSON lines to extract type fields and update per-category breakdowns.
     */
    private fun updateBreakdowns(
        stats: SyncStats,
        recordType: String,
        ndjsonPayload: String?,
        recordCount: Int
    ) {
        when (recordType) {
            "samples" -> {
                if (ndjsonPayload != null) {
                    for (type in parseNdjsonField(ndjsonPayload, "type")) {
                        stats.sampleBreakdown[type] =
                            (stats.sampleBreakdown[type] ?: 0) + 1
                    }
                }
            }
            "workouts" -> {
                if (ndjsonPayload != null) {
                    for (exerciseType in parseNdjsonField(ndjsonPayload, "exercise_type")) {
                        stats.workoutBreakdown[exerciseType] =
                            (stats.workoutBreakdown[exerciseType] ?: 0) + 1
                    }
                }
            }
            "sleep" -> {
                stats.sleepSessionCount += recordCount
            }
            "daily_summaries" -> {
                stats.dailySummaryDayCount += recordCount
            }
            "deletes" -> {
                if (ndjsonPayload != null) {
                    for (rt in parseNdjsonField(ndjsonPayload, "record_type")) {
                        stats.deleteBreakdown[rt] =
                            (stats.deleteBreakdown[rt] ?: 0) + 1
                    }
                }
            }
        }
    }

    /**
     * Splits NDJSON by newline, parses each line as a JSON object, and extracts
     * the string value of [fieldName]. Lines that fail to parse or lack the field
     * are silently skipped.
     */
    private fun parseNdjsonField(ndjson: String, fieldName: String): List<String> {
        return ndjson.split("\n")
            .filter { it.isNotBlank() }
            .mapNotNull { line ->
                try {
                    val element = Json.parseToJsonElement(line)
                    element.jsonObject[fieldName]?.jsonPrimitive?.content
                } catch (e: Exception) {
                    Log.w(TAG, "Failed to parse NDJSON line for field '$fieldName': ${e.message}")
                    null
                }
            }
    }

    // ---- File I/O Helpers ----

    private fun loadStats(): SyncStats {
        return loadFile(STATS_FILE) ?: SyncStats.empty()
    }

    private fun loadRecentEvents(): List<SyncRecord> {
        return loadFile(RECENT_EVENTS_FILE) ?: emptyList()
    }

    private inline fun <reified T> loadFile(fileName: String): T? {
        val file = File(ledgerDir, fileName)
        if (!file.exists()) return null
        return try {
            json.decodeFromString<T>(file.readText())
        } catch (e: Exception) {
            Log.w(TAG, "Failed to load $fileName: ${e.message}")
            null
        }
    }

    private fun saveFile(fileName: String, content: String) {
        try {
            File(ledgerDir, fileName).writeText(content)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to write $fileName: ${e.message}")
        }
    }
}
